pubsublite/ps/subscriber.go

```diff
@@ -0,0 +1,313 @@
+// partition will be outstanding at a time, and blocking in this receiver
+// callback will block the delivery of subsequent messages for the partition.
+type MessageReceiverFunc func(context.Context, *pubsub.Message)
+
+// SubscriberClient is a Cloud Pub/Sub Lite client to receive messages for a
+// given subscription.
+//
+// See https://cloud.google.com/pubsub/lite/docs/subscribing for more
+// information about receiving messages.
+type SubscriberClient struct {
+	settings       ReceiveSettings
+	wireSubFactory wireSubscriberFactory
+
+	// Fields below must be guarded with mu.
+	mu            sync.Mutex
+	receiveActive bool
+}
+
+// NewSubscriberClient creates a new Cloud Pub/Sub Lite client to receive
+// messages for a given subscription.
+//
+// See https://cloud.google.com/pubsub/lite/docs/subscribing for more
+// information about receiving messages.
+func NewSubscriberClient(ctx context.Context, settings ReceiveSettings, subscription pubsublite.SubscriptionPath, opts ...option.ClientOption) (*SubscriberClient, error) {
+	region, err := pubsublite.ZoneToRegion(subscription.Zone)
+	if err != nil {
+		return nil, err
+	}
+	factory := &wireSubscriberFactoryImpl{
+		settings:     settings.toWireSettings(),
+		region:       region,
+		subscription: subscription,
+		options:      opts,
+	}
+	subClient := &SubscriberClient{
+		settings:       settings,
+		wireSubFactory: factory,
+	}
+	return subClient, nil
+}
+
+// Receive calls f with the messages from the subscription. It blocks until ctx
+// is done, or the service returns a non-retryable error.
+//
+// The standard way to terminate a Receive is to cancel its context:
+//
+//   cctx, cancel := context.WithCancel(ctx)
+//   err := sub.Receive(cctx, callback)
+//   // Call cancel from callback, or another goroutine.
+//
+// If there is a fatal service error, Receive returns that error after all of
+// the outstanding calls to f have returned. If ctx is done, Receive returns nil
+// after all of the outstanding calls to f have returned and all messages have
+// been acknowledged.
+//
+// Receive calls f concurrently from multiple goroutines if the SubscriberClient
+// is connected to multiple partitions. All messages received by f must be ACKed
+// or NACKed. Failure to do so can prevent Receive from returning.
+//
+// The context passed to f will be canceled when ctx is Done or there is a fatal
+// service error.
+//
+// Each SubscriberClient may have only one invocation of Receive active at a
+// time.
+func (s *SubscriberClient) Receive(ctx context.Context, f MessageReceiverFunc) error {
+	// Initialize a subscriber instance.
+	subInstance, err := func() (*subscriberInstance, error) {
+		s.mu.Lock()
+		defer s.mu.Unlock()
+
+		if s.receiveActive {
+			return nil, errDuplicateReceive
+		}
+		subInstance, err := newSubscriberInstance(ctx, s.wireSubFactory, s.settings, f)
+		if err != nil {
+			return nil, err
+		}
+
+		s.receiveActive = true
+		return subInstance, nil
+	}()
```

Comment:
nit: I'm not 100% I see the value in wrapping this inside an anonymous function. In each of the return statements here, we should just return the error directly to `Receive` correct?

Reply:
Agree this looks a little strange. I wanted to ensure subInstance.Wait() isn't holding the mutex - mostly just so that a duplicate Receive() call returns an error immediately rather than blocking.

Refactored this to be more readable.